### DISCUSSION TITLE
ChatGPT and ingredient parsing

### DISCUSSION THREAD

**Discussion by jecorn** [mealie-recipes/mealie#2359] [General]
On Discord I've seen discussion of ChatGPT parsing of ingredients. Anecdotally, it "didn't work well", but I didn't see any specifics. So I thought to just try it out and post the results. Personally, I was alternatively pleasantly surprised and disappointed. In two tests (English recipes from the mealie demo), ChatGPT got the parsing almost perfect and also clearly recognized the recipe. But in a few other cases, it didn't really parse at all. Hopefully these 3 examples help us think a bit about NLP parsing. Note that in the responses below, it was **ChatGPT** that automatically made the tables in beautiful markdown format.

### **PROMPT**
Please separate the following list of ingredients into amounts, units, foods, and preparation notes:
28 ounces (800g) boneless/skinless chicken breast (cut into 1 inch cubes)
1 tablespoon shaoxing wine (or dry sherry)
1 tablespoon light soy sauce
2 teaspoons baking soda
1 teaspoon cornstarch / corn flour
1/2 cup low-sodium chicken stock ((or broth) -- water can be used)
5 tablespoons light soy sauce
2 tablespoons Chinese black vinegar ((or substitute good-quality balsamic vinegar))
2 tablespoon Chinese Shaoxing wine ((or dry sherry))
2 teaspoon dark soy sauce
2 teaspoons hoisin sauce
2 tablespoons sugar*
1 teaspoon cornstarch / corn flour
4 tablespoons cooking oil (divided)
1 1/2 tablespoons garlic ((4-6 cloves))
1 tablespoon ginger
1/2 red bell pepper (capsicum) (seeded and diced)
1/2 green bell pepper (capsicum) (seeded and diced)
8-10 dried chilies (cut into ½-inch pieces (adjust to taste))
1 tablespoon Sichuan peppercorns, (lightly toasted and ground**)
4 green onion / scallion stems (cut into 1-inch pieces)
1/2 cup roasted/unsalted peanuts
2 teaspoons sesame oil (optional)

### **RESPONSE**
Sure, here's a breakdown of the ingredients by amounts, units, foods, and preparation notes:

Amount | Unit | Food | Preparation notes
-------|------|------|------------------
28 | ounces (800g) | boneless/skinless chicken breast | cut into 1 inch cubes
1 | tablespoon | shaoxing wine | or dry sherry
1 | tablespoon | light soy sauce | 
2 | teaspoons | baking soda | 
1 | teaspoon | cornstarch/corn flour | 
1/2 | cup | low-sodium chicken stock | or broth; water can be used as a substitute
5 | tablespoons | light soy sauce | 
2 | tablespoons | Chinese black vinegar | or substitute with good-quality balsamic vinegar
2 | tablespoons | Chinese Shaoxing wine | or dry sherry
2 | teaspoons | dark soy sauce | 
2 | teaspoons | hoisin sauce | 
2 | tablespoons | sugar* | 
1 | teaspoon | cornstarch/corn flour | 
4 | tablespoons | cooking oil | divided
1 1/2 | tablespoons | garlic | 4-6 cloves
1 | tablespoon | ginger | 
1/2 | red bell pepper (capsicum) | seeded and diced
1/2 | green bell pepper (capsicum) | seeded and diced
8-10 | dried chilies | cut into 1/2-inch pieces (adjust to taste)
1 | tablespoon | Sichuan peppercorns | lightly toasted and ground**
4 | green onion/scallion stems | cut into 1-inch pieces
1/2 | cup | roasted/unsalted peanuts | 
2 | teaspoons | sesame oil | optional

Notes:
- For the chicken, it is recommended to use boneless and skinless chicken breast, cut into 1-inch cubes.
- Baking soda is used to tenderize the chicken.
- Cornstarch/corn flour is used to thicken the sauce.
- Low-sodium chicken stock or broth can be used, but water can be used as a substitute.
- For the dried chilies, adjust the amount according to your preferred level of spiciness.
- Sichuan peppercorns should be lightly toasted and ground before use.
- Sesame oil is optional and can be used as a finishing touch.****

### **PROMPT**
Please separate the following list of ingredients into amounts, units, foods, and preparation notes:
8 oz unsalted butter (room temp)
14 oz sugar
6 large egg whites (fresh not boxed at room temp)
14 oz AP flour
2 1/2 tsp baking powder
1/2 tsp salt
1 tsp almond extract (use clear for a whiter cake)
1 tsp vanilla extract
10 oz milk (room temp)
2 oz vegetable oil
8 oz pasteurized egg whites (room temperature)
32 oz powdered sugar
32 oz unsalted butter (room temperature)
1/2 tsp salt
1 Tbsp vanilla extract
6 oz white chocolate
1 oz warm water
1 tsp warm brown food coloring
2 tsp Truly Mad Plastics super gold
1 Tbsp Everclear (lemon extract or rose water can be used )

### **RESPONSE** 
(it's a little creepy how ChatGPT auto-separated teh ingredient list into components for each part)
Sure, here's a breakdown of the ingredients by amounts, units, foods, and preparation notes:

Cake Batter:

Amount | Unit | Food | Preparation notes
-------|------|------|------------------
8 | oz | unsalted butter | room temp
14 | oz | sugar | 
6 | large | egg whites | fresh, not boxed, at room temp
14 | oz | AP flour | 
2 1/2 | tsp | baking powder | 
1/2 | tsp | salt | 
1 | tsp | almond extract | use clear for a whiter cake
1 | tsp | vanilla extract | 
10 | oz | milk | room temp
2 | oz | vegetable oil | 

Buttercream Frosting:

Amount | Unit | Food | Preparation notes
-------|------|------|------------------
32 | oz | powdered sugar | 
32 | oz | unsalted butter | room temp
1/2 | tsp | salt | 
1 | Tbsp | vanilla extract | 

Decoration:

Amount | Unit | Food | Preparation notes
-------|------|------|------------------
6 | oz | white chocolate | 
1 | oz | warm water | 
1 | tsp | warm brown food coloring | 
2 | tsp | Truly Mad Plastics super gold | 
1 | Tbsp | Everclear | lemon extract or rose water can be used

Notes:
- For the cake batter, make sure the butter, egg whites, and milk are at room temperature before starting.
- The almond extract can be substituted with clear almond extract for a whiter cake.
- For the buttercream frosting, make sure the butter is at room temperature before starting.
- For the decoration, warm water can be used to melt the white chocolate and brown food coloring can be added to achieve the desired color. The Truly Mad Plastics super gold and Everclear (or lemon extract or rose water) can be mixed together to create a gold paint for decorating.

But not everything is cakes and ale in the land of ChatGPT:
### **PROMPT**
Please separate the following list of ingredients into amounts, units, foods, and preparation notes:
500g skinless boneless chicken thighs, cut into 3 pieces
1 lemon, juiced
2 tsp ground cumin
2 tsp paprika
1-2 tsp hot chilli powder
200g natural yogurt
2 tbsp butter, ghee or vegetable oil
1 large onion, chopped
3 garlic cloves, crushed
1 green chilli, deseeded and finely chopped (optional)
thumb-sized piece ginger, grated
1 tsp garam masala
2 tsp ground fenugreek
3 tbsp tomato purée
250ml chicken stock
50g flaked almonds, toasted
cooked basmati rice
naan bread
mango chutney or lime pickle
fresh coriander
lime wedges

### **RESPONSE**
- 500g skinless boneless chicken thighs, cut into 3 pieces
- 1 lemon, juiced
- 2 tsp ground cumin
- 2 tsp paprika
- 1-2 tsp hot chilli powder
- 200g natural yogurt
- 2 tbsp butter, ghee or vegetable oil
- 1 large onion, chopped
- 3 garlic cloves, crushed
- 1 green chilli, deseeded and finely chopped (optional)
- Thumb-sized piece of ginger, grated
- 1 tsp garam masala
- 2 tsp ground fenugreek
- 3 tbsp tomato purée
- 250ml chicken stock
- 50g flaked almonds, toasted
- Cooked basmati rice
- Naan bread
- Mango chutney or lime pickle
- Fresh coriander
- Lime wedges

Preparation Notes:
- Cut the skinless boneless chicken thighs into 3 pieces.
- Juice the lemon.
- Toast the flaked almonds.
- Cook the basmati rice.
- Optional: Deseed and finely chop the green chilli.

**Comment by carylee** [comment]
I've been very successful in using ChatGPT (and more recently the gpt-2.5 api) to parse html pages into Mealie's JSON format, then submitting that to the api.

**Reply by eikaramba** [reply]
care to maybe share your approach? i am very interested in this

**Comment by michael-genson** [comment]
There's now an open PR for this: https://github.com/mealie-recipes/mealie/pull/3581. Performance is pretty good with gpt-4o (only takes a 3-5 seconds for a recipe with ~10 ingredients)
